File: frontend/src/app/catalog/text-rules.ts
/**
 * Field rules shared with the backend (EntryValidator / TextRules). Both sides
 * trim Unicode White_Space characters and count length after trimming in
 * UTF-16 code units (JavaScript `length`, Java `String.length()`).
 */
export const LIMITS = {
  titleMaxLength: 200,
  creatorMaxLength: 200,
  descriptionMaxLength: 2000,
  yearMin: 1000,
  yearMax: 2100,
} as const;

const EDGE_WHITESPACE = /^\p{White_Space}+|\p{White_Space}+$/gu;

export function trimWhitespace(value: string | null | undefined): string {
  return (value ?? '').replace(EDGE_WHITESPACE, '');
}

/** Trimmed text, or null when blank. */
export function trimToNull(value: string | null | undefined): string | null {
  const trimmed = trimWhitespace(value);
  return trimmed === '' ? null : trimmed;
}

File: backend/src/main/java/demo/medialibrary/catalog/TextRules.java
package demo.medialibrary.catalog;

import java.util.regex.Pattern;

/**
 * Whitespace handling shared by every text input. Mirrors
 * {@code frontend/src/app/catalog/text-rules.ts}: both remove leading and
 * trailing characters with the Unicode White_Space property.
 */
public final class TextRules {

	private static final Pattern EDGE_WHITESPACE =
			Pattern.compile("^\\p{IsWhite_Space}+|\\p{IsWhite_Space}+$");

	private TextRules() {
	}

	/** Trimmed text, or {@code null} when the input is null or blank. */
	public static String trimToNull(String value) {
		if (value == null) {
			return null;
		}
		String trimmed = EDGE_WHITESPACE.matcher(value).replaceAll("");
		return trimmed.isEmpty() ? null : trimmed;
	}
}
